base/expr.jl

```diff
@@ -209,22 +222,36 @@ Give a hint to the compiler that it should not inline a function.
 
 Small functions are typically inlined automatically.
 By using `@noinline` on small functions, auto-inlining can be
-prevented. This is shown in the following example:
+prevented.
+
+`@noinline` can be applied immediately before the definition or in its function body.
 
 ```julia
-@noinline function smallfunction(x)
-    #=
-        Function Definition
-    =#
+# annotate long-form definition
+@noinline function longdef(x)
+    ...
+end
+
+# annotate short-form definition
+@noinline shortdef(x) = ...
+
+# annotate anonymous function that a `do` block creates
+f() do
+    @noinline
+    ...
 end
 ```
 
+!!! compat "Julia 1.7"
+    The usage within a function body requires at least Julia 1.7.
+
 !!! note
     If the function is trivial (for example returning a constant) it might get inlined anyway.
 """
 macro noinline(ex)
     esc(isa(ex, Expr) ? pushmeta!(ex, :noinline) : ex)
 end
+macro noinline() Expr(:meta, :noinline) end
```

Comment:
Does this mean we can delete `Core.@_noinline_meta()` also?

Reply:
Yeah, I think we can unify them. Will do that in a follow up PR though.